subsys/bluetooth/audio/vcs_client.c

```diff
@@ -0,0 +1,935 @@
+	vcs_inst->flags = 0;
+	vcs_inst->start_handle = 0;
+	vcs_inst->end_handle = 0;
+	vcs_inst->state_handle = 0;
+	vcs_inst->control_handle = 0;
+	vcs_inst->flag_handle = 0;
+	vcs_inst->vocs_inst_cnt = 0;
+	vcs_inst->aics_inst_cnt = 0;
+
+	memset(&vcs_inst->discover_params, 0, sizeof(vcs_inst->discover_params));
+
+	/* It's okay if these fail */
+	(void)bt_gatt_unsubscribe(conn, &vcs_inst->state_sub_params);
+	(void)bt_gatt_unsubscribe(conn, &vcs_inst->flag_sub_params);
+}
+
+static void bt_vcs_client_init(void)
+{
+	int i, j;
+
+	if (IS_ENABLED(CONFIG_BT_VOCS_CLIENT) &&
+	    CONFIG_BT_VCS_CLIENT_MAX_VOCS_INST > 0) {
+		for (i = 0; i < ARRAY_SIZE(vcs_insts); i++) {
+			for (j = 0; j < ARRAY_SIZE(vcs_insts[i].vocs); j++) {
+				vcs_insts[i].vocs[j] =
+					bt_vocs_client_free_instance_get();
+
+				__ASSERT(vcs_insts[i].vocs[j],
+					 "Could not allocated VOCS client instance");
+
+				bt_vocs_client_cb_register(vcs_insts[i].vocs[j],
+							   &vcs_client_cb->vocs_cb);
+			}
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_AICS_CLIENT) &&
+	    CONFIG_BT_VCS_CLIENT_MAX_AICS_INST > 0) {
+		for (i = 0; i < ARRAY_SIZE(vcs_insts); i++) {
+			for (j = 0; j < ARRAY_SIZE(vcs_insts[i].aics); j++) {
+				vcs_insts[i].aics[j] =
+					bt_aics_client_free_instance_get();
+
+				__ASSERT(vcs_insts[i].aics[j],
+					"Could not allocated VOCS client instance");
+
+				bt_aics_client_cb_register(vcs_insts[i].aics[j],
+							   &vcs_client_cb->aics_cb);
+			}
+		}
+	}
+}
+
+int bt_vcs_discover(struct bt_conn *conn)
+{
+	static bool initialized;
+	struct vcs_instance *vcs_inst = &vcs_insts[bt_conn_index(conn)];
+
+	/*
+	 * This will initiate a discover procedure. The procedure will do the
+	 * following sequence:
+	 * 1) Primary discover for the VCS
+	 * 2) Characteristic discover of the VCS
+	 * 3) Discover services included in VCS (VOCS and AICS)
+	 * 4) For each included service found; discovery of the characteristics
+	 * 5) When everything above have been discovered, the callback is called
+	 */
+
+	if (!conn) {
+		return -ENOTCONN;
+	} else if (vcs_inst->busy) {
+		return -EBUSY;
+	}
+
+	if (!initialized) {
+		bt_vcs_client_init();
+		initialized = true;
+	}
+
+	vcs_client_reset(conn);
+
+	memcpy(&vcs_inst->uuid, BT_UUID_VCS, sizeof(vcs_inst->uuid));
+
+	vcs_inst->discover_params.func = primary_discover_func;
+	vcs_inst->discover_params.uuid = &vcs_inst->uuid.uuid;
+	vcs_inst->discover_params.type = BT_GATT_DISCOVER_PRIMARY;
+	vcs_inst->discover_params.start_handle = BT_ATT_FIRST_ATTTRIBUTE_HANDLE;
+	vcs_inst->discover_params.end_handle = BT_ATT_LAST_ATTTRIBUTE_HANDLE;
+
+	return bt_gatt_discover(conn, &vcs_inst->discover_params);
+}
+
+void bt_vcs_client_cb_register(struct bt_vcs_cb *cb)
+{
+	int i, j;
+
+	vcs_client_cb = cb;
+
+	if (IS_ENABLED(CONFIG_BT_VOCS_CLIENT)) {
+		struct bt_vocs_cb *vocs_cb = NULL;
+
+		if (cb) {
+			if (cb->vocs_cb.discover) {
+				BT_WARN("VCS overwrote discover callback of VOCS");
+			}
```

Comment:
I needed to think on this one.
The VOCS discover callback is overwritten (or set) in all cases - whether it was set before or not.
But the _warning_ is only given if it is set before.

Reply:
I thought that made sense; if it was not set, then it was not really overwritten :)

Reply:
Agreed - it is absolutely correct.  But I somehow came from a different angle.
Maybe it could be augmented with an 
else
        BT_DEBUG("Set the discover callback of VOCS")?
Or possibly a comment of "/* Set VOCS callback - overwrite if already set */ ?

Reply:
Actually, the better solution here might be to simply not allow the application to set the callback (can be verified in the register function), assuming that the application does not change the callback structure after registering (but that assumption goes for basically all callbacks in Zephyr). 

@asbjornsabo Do you think that makes more sense?

Reply:
I am starting to feel that I am making a lot of noise about a little thing ...
The thing works, and is correct, as it is, so although I was confused there is no pressing need for a change.
But I can not see now why the application would want to supply it's own callback, so your proposal makes sense.
Do as you think is best.